File: proxy/models/tv_show.py
from typing import Optional, Dict, List
from dataclasses import dataclass, field
from proxy.models.base import Images, Platform, Author


@dataclass
class Episode:
    id: int
    episode_number: int
    season_number: int
    title: str
    description: Optional[str] = None
    release_date: Optional[str] = None
    duration_minutes: Optional[int] = None
    image_url: Optional[str] = None
    episode_type: Optional[str] = None

    def to_dict(self) -> Dict:
        return {
            'id': self.id,
            'season_number': self.season_number,
            'episode_number': self.episode_number,
            'title': self.title,
            'description': self.description,
            'image_url': self.image_url,
            'release_date': self.release_date,
            'duration_minutes': self.duration_minutes
        }


@dataclass
class Season:
    id: int
    season_number: int
    title: str
    content_type: str
    number_of_episodes: int = 0
    description: Optional[str] = None
    release_date: Optional[str] = None
    image_url: Optional[str] = None
    tv_show_name: Optional[str] = None
    images: Optional[Images] = None
    episodes: List[Episode] = field(default_factory=list)
    platforms: Optional[Dict[str, List[Platform]]] = None

    def to_dict(self, images_size: int = 18) -> Dict:
        result = {
            'id': self.id,
            'type': self.content_type,
            'season_number': self.season_number,
            'title': self.title,
            'tv_show_name': self.tv_show_name,
            'description': self.description,
            'image_url': self.image_url,
            'release_date': self.release_date,
            'number_of_episodes': self.number_of_episodes
        }

        if self.images:
            result['images'] = self.images.to_dict(images_size=images_size)

        if self.platforms:
            result['platforms'] = {
                country: [p.to_dict() for p in platforms]
                for country, platforms in self.platforms.items()
            }

        if self.episodes:
            result['episodes'] = [ep.to_dict() for ep in self.episodes]

        return result


@dataclass
class TVShow:
    id: int
    title: str
    original_title: str
    content_type: str
    description: Optional[str] = None
    image_url: Optional[str] = None
    tagline: Optional[str] = None
    imdb_id: Optional[str] = None
    release_date: Optional[str] = None
    status: Optional[str] = None
    number_of_seasons: Optional[int] = None
    number_of_episodes: Optional[int] = None
    authors: Optional[List[Author]] = None
    images: Optional[Images] = None
    platforms: Optional[Dict[str, List[Platform]]] = None
    seasons: List[Season] = field(default_factory=list)

    def to_dict(self, images_size: int = 18) -> Dict:
        result = {
            'id': self.id,
            'type': self.content_type,
            'imdb_id': self.imdb_id,
            'title': self.title,
            'original_title': self.original_title,
            'tagline': self.tagline,
            'description': self.description,
            'image_url': self.image_url,
            'release_date': self.release_date,
            'status': self.status,
            'number_of_seasons': self.number_of_seasons,
            'number_of_episodes': self.number_of_episodes
        }

        if self.authors:
            result['authors'] = [author.to_dict() for author in self.authors]

        if self.images:
            result['images'] = self.images.to_dict(images_size=images_size)

        if self.platforms:
            result['platforms'] = {
                country: [platform.to_dict() for platform in platforms]
                for country, platforms in self.platforms.items()
            }

        if self.seasons:
            result['seasons'] = [season.to_dict(images_size=images_size) for season in self.seasons]
        return result

File: internal/models/tvshow.go
package models

type Episode struct {
	ID              string  `json:"id"`
	EpisodeNumber   int     `json:"episode_number"`
	SeasonNumber    int     `json:"season_number"`
	Title           string  `json:"title"`
	Description     *string `json:"description"`
	ReleaseDate     *string `json:"release_date"`
	DurationMinutes *int    `json:"duration_minutes"`
	ImageURL        *string `json:"image_url"`
	EpisodeType     *string `json:"episode_type,omitempty"`
}

type Season struct {
	ID               string                `json:"id"`
	SeasonNumber     int                   `json:"season_number"`
	Title            string                `json:"title"`
	ContentType      string                `json:"type"`
	NumberOfEpisodes int                   `json:"number_of_episodes"`
	Description      *string               `json:"description"`
	ReleaseDate      *string               `json:"release_date"`
	ImageURL         *string               `json:"image_url"`
	TVShowName       *string               `json:"tv_show_name"`
	Images           *Images               `json:"-"`
	Episodes         []Episode             `json:"-"`
	Platforms        map[string][]Platform `json:"-"`
}

type SeasonResponse struct {
	ID               string                `json:"id"`
	Type             string                `json:"type"`
	SeasonNumber     int                   `json:"season_number"`
	Title            string                `json:"title"`
	TVShowName       *string               `json:"tv_show_name"`
	Description      *string               `json:"description"`
	ImageURL         *string               `json:"image_url"`
	ReleaseDate      *string               `json:"release_date"`
	NumberOfEpisodes int                   `json:"number_of_episodes"`
	Images           []ImageEntry          `json:"images,omitempty"`
	Platforms        map[string][]Platform `json:"platforms,omitempty"`
	Episodes         []Episode             `json:"episodes,omitempty"`
}

func (s *Season) ToResponse(imagesSize int) SeasonResponse {
	resp := SeasonResponse{
		ID:               s.ID,
		Type:             s.ContentType,
		SeasonNumber:     s.SeasonNumber,
		Title:            s.Title,
		TVShowName:       s.TVShowName,
		Description:      s.Description,
		ImageURL:         s.ImageURL,
		ReleaseDate:      s.ReleaseDate,
		NumberOfEpisodes: s.NumberOfEpisodes,
		Episodes:         s.Episodes,
	}

	if s.Images != nil {
		resp.Images = s.Images.ToList(imagesSize)
	}

	if s.Platforms != nil {
		resp.Platforms = s.Platforms
	}

	return resp
}

type TVShow struct {
	ID               string                `json:"id"`
	Title            string                `json:"title"`
	OriginalTitle    string                `json:"original_title"`
	ContentType      string                `json:"type"`
	Description      *string               `json:"description"`
	ImageURL         *string               `json:"image_url"`
	Tagline          *string               `json:"tagline"`
	ImdbID           *string               `json:"imdb_id"`
	ReleaseDate      *string               `json:"release_date"`
	Status           *string               `json:"status"`
	NumberOfSeasons  *int                  `json:"number_of_seasons"`
	NumberOfEpisodes *int                  `json:"number_of_episodes"`
	Authors          []Author              `json:"authors,omitempty"`
	Images           *Images               `json:"-"`
	Platforms        map[string][]Platform `json:"-"`
	Seasons          []Season              `json:"-"`
}

type TVShowResponse struct {
	ID               string                `json:"id"`
	Type             string                `json:"type"`
	ImdbID           *string               `json:"imdb_id"`
	Title            string                `json:"title"`
	OriginalTitle    string                `json:"original_title"`
	Tagline          *string               `json:"tagline"`
	Description      *string               `json:"description"`
	ImageURL         *string               `json:"image_url"`
	ReleaseDate      *string               `json:"release_date"`
	Status           *string               `json:"status"`
	NumberOfSeasons  *int                  `json:"number_of_seasons"`
	NumberOfEpisodes *int                  `json:"number_of_episodes"`
	Authors          []Author              `json:"authors,omitempty"`
	Images           []ImageEntry          `json:"images,omitempty"`
	Platforms        map[string][]Platform `json:"platforms,omitempty"`
	Seasons          []SeasonResponse      `json:"seasons,omitempty"`
}

func (tv *TVShow) ToResponse(imagesSize int) TVShowResponse {
	resp := TVShowResponse{
		ID:               tv.ID,
		Type:             tv.ContentType,
		ImdbID:           tv.ImdbID,
		Title:            tv.Title,
		OriginalTitle:    tv.OriginalTitle,
		Tagline:          tv.Tagline,
		Description:      tv.Description,
		ImageURL:         tv.ImageURL,
		ReleaseDate:      tv.ReleaseDate,
		Status:           tv.Status,
		NumberOfSeasons:  tv.NumberOfSeasons,
		NumberOfEpisodes: tv.NumberOfEpisodes,
		Authors:          tv.Authors,
	}

	if tv.Images != nil {
		resp.Images = tv.Images.ToList(imagesSize)
	}

	if tv.Platforms != nil {
		resp.Platforms = tv.Platforms
	}

	if len(tv.Seasons) > 0 {
		resp.Seasons = make([]SeasonResponse, len(tv.Seasons))

		for i, season := range tv.Seasons {
			resp.Seasons[i] = season.ToResponse(imagesSize)
		}
	}

	return resp
}
